src/Build/Evaluation/LazyItemEvaluator.cs

```diff
@@ -219,17 +219,299 @@ public void MarkAsReferenced()
+                /// A dictionary of items keyed by their normalized value.
+                /// </summary>
+                private ImmutableDictionary<string, DictionaryValue>.Builder _dictionaryBuilder;
+
+                internal Builder(int version, ImmutableList<ItemData>.Builder listBuilder, ImmutableDictionary<string, DictionaryValue>.Builder dictionaryBuilder)
+                {
+                    _version = version;
+                    _listBuilder = listBuilder;
+                    _dictionaryBuilder = dictionaryBuilder;
+                }
+
+                #region IEnumerable implementation
+
+                public IEnumerator<ItemData> GetEnumerator() => _listBuilder.GetEnumerator();
+
+                System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => _listBuilder.GetEnumerator();
+
+                #endregion
+
+                public int Count => _listBuilder.Count;
+
+                public ItemData this[int index]
+                {
+                    get { return _listBuilder[index]; }
+                    set
+                    {
+                        _listBuilder[index] = value;
+                        // This is a rare operation, don't bother updating the dictionary for now. It will be recreated as needed.
+                        _dictionaryBuilder = null;
+                    }
+                }
+
+                public void Add(ItemData data)
+                {
+                    _listBuilder.Add(data);
+                    if (_dictionaryBuilder != null)
+                    {
+                        AddToDictionary(data.Item);
+                    }
+                }
+
+                public void Clear()
+                {
+                    _listBuilder.Clear();
+                    _dictionaryBuilder?.Clear();
+                }
+
+                public void RemoveAll(ICollection<I> itemsToRemove)
+                {
+                    _listBuilder.RemoveAll(item => itemsToRemove.Contains(item.Item));
+                    // This is a rare operation, don't bother updating the dictionary for now. It will be recreated as needed.
+                    _dictionaryBuilder = null;
+                }
+
+                /// <summary>
+                /// Removes items from the collection that match the given ItemSpec.
+                /// </summary>
+                /// <remarks>
+                /// If <see cref="_dictionaryBuilder"/> does not exist yet, it is created in this method to avoid the cost of comparing each item
+                /// being removed with each item already in the collection. The dictionary is kept in sync with the <see cref="_listBuilder"/>
+                /// as long as practical. If an operation would result in too much of such work, the dictionary is simply dropped and recreated
+                /// later if/when needed.
+                /// </remarks>
+                public void RemoveMatchingItems(ItemSpec<P, I> itemSpec)
```

Comment:
Wouldn't it be better design-wise to ask the ItemSpec to compute its intersection with a given OrderedItemDataCollection via a new `IntersectWith(OrderedItemDataCollection)`?

That way you:
- prevent leaking the implementation detail that some itemspec fragments do not support enumeration 
- can neatly cache normalized values of the itemspec values without leaking that implementation detail. This would be useful if itemspec values get repeatedly normalized (for example by repeatedly calling any of the IsMatch, MatchCount, or IntersectWith methods)
- can reuse the IntersectWith operation to remove the same O(n^2) issue with the Update operation.

To support this, OrderedItemDataCollection (or the Builder) could provide a string indexer which would be the normalized key. Ideally you could even hide OrderedItemDataCollection behind some collection interface but maybe that's too much for now.

Reply:
This is a great suggestion. I think I had the code structured differently but hit a technical issue which had me settle for this mess. I'll wait for CI to start working again to see where I stand with unit tests and will then tackle this. Thank you!

Reply:
I have refactored it a bit as suggested and it is indeed nicer. For Update operation, it is possible that we can now revert the previous optimization (#5853) and achieve the same using the new datastructure but I'd like to try this separately later.